### DISCUSSION TITLE
Grafana azure sso login is always giving Viewer role even though user is assigned to Admin and Editor role groups?

### DISCUSSION THREAD

**Discussion by SudeshnaBora** [grafana/grafana#70520] [General]
I have a grafana application registered in azure AD. The appRole mentioned in the manifest is assigned to azure groups I have created. To run the grafana application locally, I have the following config and docker file.

The grafana configuration is as follows :

```
[auth.azuread]
name = Azure AD
enabled = true
allow_sign_up = true
auto_login = true
client_id = <client_id>
client_secret = <client_secret>
scopes = openid email profile
auth_url = https://login.microsoftonline.com/<tenant_id>/oauth2/v2.0/authorize
token_url = https://login.microsoftonline.com/<tenant_id>/oauth2/v2.0/token
allowed_domains =
allowed_groups = <group_id1> <group_id2>
role_attribute_strict = false
allow_assign_grafana_admin = true
skip_org_role_sync = true
force_use_graph_api = true
[log]
level = "debug"
```

The docker compose is as follows :

```
version: '3.9'
services:
  grafana:
    image: grafana/grafana:9.4.7-ubuntu
    ports:
      - 3000:3000
    restart: unless-stopped
    volumes:
      - ./grafana_config.ini:/etc/grafana/grafana.ini
      - grafana-data:/var/lib/grafana

volumes:
  grafana-data:
```
While trying to log in , I am always logged into as Viewer role even though I am Editor.

Onwards to debugging,

I first tried to check if the correct roles and groups are returned by the access token.

While trying to get the access token , I realised that `azure access_token` do not give roles and groups. Only id_token gives us the roles and groups.

This I verified as follows :

After getting the authorization code following this tutorial https://learn.microsoft.com/en-us/azure/databricks/dev-tools/app-aad-token#interactive . i used the following curl command

```
curl -X POST -H 'Content-Type: application/x-www-form-urlencoded' \
https://login.microsoftonline.com/<tenant_id>/oauth2/v2.0/token \
-d 'client_id=<client_id>' \
-d 'scope=00000003-0000-0000-c000-000000000000%2F.default' \
-d 'code=<authorization_code>' \
-d 'redirect_uri=http://localhost' \
-d 'grant_type=authorization_code' \
-d 'state=12345' \
-d 'client_secret=<secret>'
```

Here, if I give scope as only the id of microsoft graph api , I get access_token , if I give openid as well I get the id_token.

My id token has both the roles and the groups.

This led to me trying to intercept the access token from the ui , unfortunately I do not seem to find the header containing the authorization token.

The max I get from networks tab is :

```
curl 'http://localhost:3000/login/azuread?code=<authorization code>' \
  -H 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7' \
  -H 'Accept-Language: en-US,en;q=0.9' \
  -H 'Cache-Control: max-age=0' \
  -H 'Connection: keep-alive' \
  -H 'Cookie: oauth_state=142d92d62f5fed47cf1232a8d2f92597c2e58b5414043604822ee29e8e2712d1' \
  -H 'Sec-Fetch-Dest: document' \
  -H 'Sec-Fetch-Mode: navigate' \
  -H 'Sec-Fetch-Site: cross-site' \
  -H 'Sec-Fetch-User: ?1' \
  -H 'Upgrade-Insecure-Requests: 1' \
  -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36' \
  -H 'sec-ch-ua: "Not.A/Brand";v="8", "Chromium";v="114", "Google Chrome";v="114"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "macOS"' \
  --compressed

```

I have also tried intercepting the access token but no success.

So I have the following questions :

While trying to log in via ui , does grafana application request for access token or id token or both?
In case, it is giving only access token. Is it cause the groups and roles are not part of the access token, do I get the default Viewer role mapped.
If possible, how do I intercept access token.
Finally, is there another way I can try to get the given roles or groups in the configuration files.
Ultimately, my issue is that even though I am able to log in via SSO azure . I never get any other roles.

**Comment by SudeshnaBora** [comment]
Tutorial source : https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/azuread/

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82663. Feel free to continue the discussion around this there. Thank you!
